File: python/maximum_sum_circular_subarray.py
class Solution:
    def maxSubarraySumCircular(self, nums: list[int]) -> int:
        maxSum: int = nums[0]
        currentMaxSum: int = 0
        minSum: int = nums[0]
        currentMinSum: int = 0
        total: int = 0
        for i in range(len(nums)):
            total += nums[i]
            currentMaxSum = max(nums[i], currentMaxSum + nums[i])
            maxSum = max(currentMaxSum, maxSum)
            currentMinSum = min(nums[i], currentMinSum + nums[i])
            minSum = min(currentMinSum, minSum)

        return max(maxSum, total - minSum) if maxSum > 0 else maxSum


def test_maxSubarraySumCircular_case_1():
    # arrange
    nums: list[int] = [1, -2, 3, -2]
    expected: int = 3

    # act
    solution = Solution()
    actual = solution.maxSubarraySumCircular(nums)

    # assert
    assert expected == actual


def test_maxSubarraySumCircular_case_2():
    # arrange
    nums: list[int] = [5, -3, 5]
    expected: int = 10

    # act
    solution = Solution()
    actual = solution.maxSubarraySumCircular(nums)

    # assert
    assert expected == actual


def test_maxSubarraySumCircular_case_3():
    # arrange
    nums: list[int] = [-3, -2, -3]
    expected: int = -2

    # act
    solution = Solution()
    actual = solution.maxSubarraySumCircular(nums)

    # assert
    assert expected == actual


def test_maxSubarraySumCircular_case_4():
    # arrange
    nums: list[int] = [-2, -3, -1]
    expected: int = -1

    # act
    solution = Solution()
    actual = solution.maxSubarraySumCircular(nums)

    # assert
    assert expected == actual


def test_maxSubarraySumCircular_case_5():
    # arrange
    nums: list[int] = [-5, 3, 5]
    expected: int = 8

    # act
    solution = Solution()
    actual = solution.maxSubarraySumCircular(nums)

    # assert
    assert expected == actual


def test_maxSubarraySumCircular_case_6():
    # arrange
    nums: list[int] = [1, -6, -7, 4]
    expected: int = 5

    # act
    solution = Solution()
    actual = solution.maxSubarraySumCircular(nums)

    # assert
    assert expected == actual

File: rust/src/code/maximum_sum_circular_subarray.rs
struct Solution {}

impl Solution {
    pub fn max_subarray_sum_circular(nums: Vec<i32>) -> i32 {
        let mut max_sum: i32 = nums[0];
        let mut max_curr_sum: i32 = 0;
        let mut min_sum: i32 = nums[0];
        let mut min_curr_sum: i32 = 0;
        let mut total: i32 = 0;
        for i in 0..nums.len() {
            total += nums[i];
            max_curr_sum = std::cmp::max(nums[i], max_curr_sum + nums[i]);
            min_curr_sum = std::cmp::min(nums[i], min_curr_sum + nums[i]);
            max_sum = std::cmp::max(max_sum, max_curr_sum);
            min_sum = std::cmp::min(min_sum, min_curr_sum);
        }

        return if max_sum > 0 {
            std::cmp::max(total - min_sum, max_sum)
        } else {
            max_sum
        };
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_max_subarray_sum_circular() {
        // arrange
        let nums: Vec<i32> = vec![1, -2, 3, -2];
        let expected: i32 = 3;

        // act
        let actual = Solution::max_subarray_sum_circular(nums);

        // assert
        assert_eq!(expected, actual);
    }
}
